tmt/cli.py

```diff
@@ -106,12 +108,25 @@ class CliInvocation:
     def from_context(cls, context: Context) -> 'CliInvocation':
         return CliInvocation(context=context, options=context.params)
 
+    @classmethod
+    def from_options(cls, options: Dict[str, Any]) -> 'CliInvocation':
+        """ Inject custom options coming from the command line """
+        invocation = CliInvocation(context=None, options=options)
+        invocation._sources = {
+            key: click.core.ParameterSource.COMMANDLINE
+            for key in options.keys()
+            }
+        return invocation
+
     @cached_property
     def option_sources(self) -> Dict[str, click.core.ParameterSource]:
-        if not self.context:
-            return {}
+        try:
+            return self._sources
+        except AttributeError:
+            if not self.context:
+                return {}
 
-        return self.context._parameter_source
+            return self.context._parameter_source
```

Comment:
I'd like to leave `option_sources` as it was. IIRC, the `cached_property` shipped with the standard library supports setting the cached property, treating it as an attribute, which then disabled the dynamic nature. I'll try to find out how to extend our much simpler implementation, but eventually, this function would stay as it is, and `from_options()` would do something like `invocation.option_sources = {...}`.

Reply:
The thing is that for the use case of injecting just `image` and `how` to the provision using

```python
Provision.store_cli_invocation(context=None, options=options)
```

there is no provision `context` to be handed over. So my idea was to provide a way to only inject options. I'm open to any idea how to cover this better. With the current approach `mypy` is unhappy as well:

```
tmt/cli.py:115: error: "CliInvocation" has no attribute "_sources"  [attr-defined]
tmt/cli.py:124: error: Returning Any from function declared to return "dict[str, ParameterSource]"  [no-any-return]
tmt/cli.py:124: error: "CliInvocation" has no attribute "_sources"  [attr-defined]
```

And adding `_sources` to the `CliInvocation` resulted in a different bunch of errors. @happz, do you have some elegant solution up your sleeve?

Reply:
I have an idea, I'll give it a few tries.

Reply:
If you still have the idea that would be great ;) Or at least a hint how to make `mypy` happy would help me.

Reply:
Ok, for now, I've made `mypy` happy by c7c01b05. We can simplify later.